File: scripts/study.py
#!/usr/bin/env python3
"""Small Markdown automation helpers for this study-route repository."""

from __future__ import annotations

import argparse
import datetime as dt
from pathlib import Path
from typing import Iterable

from study_route_paths import DATA_ROOT

DEFAULT_GOAL = "goals/2026-internship-ai-backend.md"
DEFAULT_PLAN = "plans/2026-W25-project-kickoff.md"


def today() -> str:
    return dt.date.today().isoformat()


def read_text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def write_text(path: Path, text: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text, encoding="utf-8", newline="\n")


def parse_row(value: str, fields: int, label: str) -> list[str]:
    parts = [part.strip() for part in value.split("|")]
    if len(parts) != fields:
        raise SystemExit(f"{label} 需要 {fields} 个字段，用 | 分隔：{value}")
    return parts


def markdown_row(cells: Iterable[str]) -> str:
    cleaned = [cell.replace("\n", " ").strip() for cell in cells]
    return "| " + " | ".join(cleaned) + " |"


def log_path(date: str) -> Path:
    return DATA_ROOT / "logs" / f"{date}.md"


def render_log(date: str, status: str, hours: str, goal: str, plan: str) -> str:
    return f"""# 学习记录

## 基本信息

- 日期：{date}
- 关联目标：`{goal}`
- 关联计划：`{plan}`
- 实际学习时长：{hours}
- 今日状态：{status}

## 今日完成

| 任务 | 结果 | 用时 | 证据或产出 |
| --- | --- | ---: | --- |

## 关键收获

## 遇到的问题

| 问题 | 当前判断 | 下一步 |
| --- | --- | --- |

## 明日计划

## 备注
"""


def ensure_log(date: str, status: str, hours: str, goal: str, plan: str) -> Path:
    path = log_path(date)
    if not path.exists():
        write_text(path, render_log(date, status, hours, goal, plan))
    return path


def append_after_table(text: str, heading: str, rows: list[str]) -> str:
    if not rows:
        return text
    marker = f"## {heading}"
    start = text.find(marker)
    if start == -1:
        raise SystemExit(f"找不到章节：{marker}")
    next_heading = text.find("\n## ", start + len(marker))
    section_end = len(text) if next_heading == -1 else next_heading
    section = text[start:section_end]
    lines = section.splitlines()
    insert_at = len(lines)
    for index, line in enumerate(lines):
        if line.startswith("| ---"):
            insert_at = index + 1
            while insert_at < len(lines) and lines[insert_at].startswith("|"):
                insert_at += 1
            break
    for row in reversed(rows):
        lines.insert(insert_at, row)
    new_section = "\n".join(lines)
    return text[:start] + new_section + text[section_end:]


def append_bullets(text: str, heading: str, items: list[str]) -> str:
    if not items:
        return text
    marker = f"## {heading}"
    start = text.find(marker)
    if start == -1:
        raise SystemExit(f"找不到章节：{marker}")
    next_heading = text.find("\n## ", start + len(marker))
    section_end = len(text) if next_heading == -1 else next_heading
    insert = "".join(f"\n- {item.strip()}" for item in items if item.strip())
    return text[:section_end].rstrip() + insert + "\n" + text[section_end:]


def cmd_init_log(args: argparse.Namespace) -> None:
    date = args.date or today()
    path = ensure_log(date, args.status, args.hours, args.goal, args.plan)
    print(path.relative_to(DATA_ROOT))


def cmd_add_log(args: argparse.Namespace) -> None:
    date = args.date or today()
    path = ensure_log(date, args.status, args.hours, args.goal, args.plan)
    text = read_text(path)

    done_rows = [markdown_row(parse_row(item, 4, "--done")) for item in args.done]
    problem_rows = [markdown_row(parse_row(item, 3, "--problem")) for item in args.problem]

    text = append_after_table(text, "今日完成", done_rows)
    text = append_after_table(text, "遇到的问题", problem_rows)
    text = append_bullets(text, "关键收获", args.takeaway)
    text = append_bullets(text, "明日计划", args.next)
    text = append_bullets(text, "备注", args.note)

    write_text(path, text)
    print(path.relative_to(DATA_ROOT))


def ensure_leetcode_record() -> Path:
    path = DATA_ROOT / "records" / "leetcode.md"
    if not path.exists():
        write_text(
            path,
            """# LeetCode 刷题记录

| 日期 | 专题 | 题目 | 难度 | 结果 | 是否重做 | 备注 |
| --- | --- | --- | --- | --- | --- | --- |
""",
        )
    return path


def cmd_leetcode(args: argparse.Namespace) -> None:
    path = ensure_leetcode_record()
    date = args.date or today()
    row = markdown_row(
        [
            date,
            args.topic,
            args.title,
            args.difficulty,
            args.result,
            args.redo,
            args.note,
        ]
    )
    text = read_text(path).rstrip() + "\n" + row + "\n"
    write_text(path, text)
    print(path.relative_to(DATA_ROOT))


def ensure_exam_review_record() -> Path:
    path = DATA_ROOT / "records" / "exam-review.md"
    if not path.exists():
        write_text(
            path,
            """# 期末复习记录

| 日期 | 科目 | 计划时长 | 实际时长 | 完成内容 | 问题 | 下一步 |
| --- | --- | ---: | ---: | --- | --- | --- |
""",
        )
    return path


def cmd_exam_review(args: argparse.Namespace) -> None:
    path = ensure_exam_review_record()
    date = args.date or today()
    row = markdown_row(
        [
            date,
            args.subject,
            args.planned,
            args.actual,
            args.done,
            args.problem,
            args.next,
        ]
    )
    text = read_text(path).rstrip() + "\n" + row + "\n"
    write_text(path, text)
    print(path.relative_to(DATA_ROOT))


def replace_focus_line(text: str, prefix: str, value: str | None) -> str:
    if not value:
        return text
    lines = text.splitlines()
    target = f"- {prefix}："
    for index, line in enumerate(lines):
        if line.startswith(target):
            lines[index] = f"{target}{value}"
            return "\n".join(lines) + "\n"
    raise SystemExit(f"dashboard.md 中找不到行：{target}")


def update_progress_row(text: str, spec: str, date: str) -> str:
    project, status, stage, milestone = parse_row(spec, 4, "--progress")
    lines = text.splitlines()
    for index, line in enumerate(lines):
        if line.startswith(f"| {project} |"):
            lines[index] = markdown_row([project, status, stage, milestone, date])
            return "\n".join(lines) + "\n"
    marker = "| --- | --- | --- | --- | --- |"
    for index, line in enumerate(lines):
        if line == marker:
            lines.insert(index + 1, markdown_row([project, status, stage, milestone, date]))
            return "\n".join(lines) + "\n"
    raise SystemExit("dashboard.md 中找不到进度概览表")


def cmd_dashboard(args: argparse.Namespace) -> None:
    path = DATA_ROOT / "dashboard.md"
    text = read_text(path)
    date = args.date or today()
    text = replace_focus_line(text, "当前阶段", args.stage)
    text = replace_focus_line(text, "本周重点", args.week)
    text = replace_focus_line(text, "今日任务", args.today)
    for item in args.progress:
        text = update_progress_row(text, item, date)
    write_text(path, text)
    print(path.relative_to(DATA_ROOT))


def cmd_week_plan(args: argparse.Namespace) -> None:
    week = args.week or f"{dt.date.today().isocalendar().year}-W{dt.date.today().isocalendar().week:02d}"
    path = DATA_ROOT / "plans" / f"{week}.md"
    if path.exists() and not args.force:
        raise SystemExit(f"{path.relative_to(DATA_ROOT)} 已存在；如需覆盖请加 --force")
    text = f"""# 学习计划：{week}

## 基本信息

- 计划周期：{week}
- 关联目标：`{args.goal}`
- 关联路线：`{args.route}`
- 本周期主题：{args.theme}
- 计划总时长：{args.hours}

## 本周期目标

| 目标 | 验收方式 | 优先级 |
| --- | --- | --- |

## 任务安排

| 日期 | 任务 | 预计用时 | 产出物 | 状态 |
| --- | --- | ---: | --- | --- |

## 每日最低动作

- 刷 1 道算法题，并记录解法要点。
- 学习 30 分钟后端、计算机基础或 Linux 实操。
- 在 `logs/` 中记录当天完成内容、问题和下一步。

## 重点问题

| 问题 | 解决路径 | 截止时间 |
| --- | --- | --- |

## 复盘入口

- 对应复盘文件：`reviews/{week}.md`
"""
    write_text(path, text)
    print(path.relative_to(DATA_ROOT))


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="学习路线 Markdown 自动化工具")
    sub = parser.add_subparsers(dest="command", required=True)

    init_log = sub.add_parser("init-log", help="创建某天学习日志")
    init_log.add_argument("--date")
    init_log.add_argument("--status", default="顺利")
    init_log.add_argument("--hours", default="")
    init_log.add_argument("--goal", default=DEFAULT_GOAL)
    init_log.add_argument("--plan", default=DEFAULT_PLAN)
    init_log.set_defaults(func=cmd_init_log)

    add_log = sub.add_parser("add-log", help="向某天学习日志追加内容")
    add_log.add_argument("--date")
    add_log.add_argument("--status", default="顺利")
    add_log.add_argument("--hours", default="")
    add_log.add_argument("--goal", default=DEFAULT_GOAL)
    add_log.add_argument("--plan", default=DEFAULT_PLAN)
    add_log.add_argument("--done", action="append", default=[], help="任务|结果|用时|证据或产出")
    add_log.add_argument("--takeaway", action="append", default=[])
    add_log.add_argument("--problem", action="append", default=[], help="问题|当前判断|下一步")
    add_log.add_argument("--next", action="append", default=[])
    add_log.add_argument("--note", action="append", default=[])
    add_log.set_defaults(func=cmd_add_log)

    leetcode = sub.add_parser("leetcode", help="追加 LeetCode 刷题记录")
    leetcode.add_argument("--date")
    leetcode.add_argument("--topic", required=True)
    leetcode.add_argument("--title", required=True)
    leetcode.add_argument("--difficulty", default="中等")
    leetcode.add_argument("--result", default="待复盘")
    leetcode.add_argument("--redo", default="否")
    leetcode.add_argument("--note", default="")
    leetcode.set_defaults(func=cmd_leetcode)

    exam_review = sub.add_parser("exam-review", help="追加期末复习时长记录")
    exam_review.add_argument("--date")
    exam_review.add_argument("--subject", required=True)
    exam_review.add_argument("--planned", default="")
    exam_review.add_argument("--actual", required=True)
    exam_review.add_argument("--done", default="")
    exam_review.add_argument("--problem", default="")
    exam_review.add_argument("--next", default="")
    exam_review.set_defaults(func=cmd_exam_review)

    dashboard = sub.add_parser("dashboard", help="更新 dashboard 当前焦点或进度概览")
    dashboard.add_argument("--date")
    dashboard.add_argument("--stage")
    dashboard.add_argument("--week")
    dashboard.add_argument("--today")
    dashboard.add_argument("--progress", action="append", default=[], help="项目|状态|当前阶段|下一个里程碑")
    dashboard.set_defaults(func=cmd_dashboard)

    week_plan = sub.add_parser("week-plan", help="创建周计划 Markdown")
    week_plan.add_argument("--week")
    week_plan.add_argument("--theme", default="待填写")
    week_plan.add_argument("--hours", default="待填写")
    week_plan.add_argument("--goal", default=DEFAULT_GOAL)
    week_plan.add_argument("--route", default="routes/backend-plus-ai-application-roadmap.md")
    week_plan.add_argument("--force", action="store_true")
    week_plan.set_defaults(func=cmd_week_plan)

    return parser


def main() -> None:
    parser = build_parser()
    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: src/cli/study.ts
#!/usr/bin/env node
import path from "node:path";
import { resolveDataConfig } from "../backend/config.js";
import { ensureRecommendedFrontMatter } from "../backend/documentModel.js";
import { defaultContent } from "../backend/templates.js";
import {
  appendAfterTable,
  appendBullet,
  markdownRow,
  readText,
  relativePath,
  writeText
} from "../backend/markdownStore.js";
import { doctorReport, healthReport, migrateWorkspace } from "../backend/workspace.js";
import type { CliCommand, ParsedCli } from "../../types/cli.js";

const DEFAULT_GOAL = "goals/2026-internship-ai-backend.md";
const DEFAULT_PLAN = "plans/2026-W25-project-kickoff.md";

function today(): string {
  return new Date().toISOString().slice(0, 10);
}

function parseArgs(argv: string[]): ParsedCli {
  const command = argv[0] as CliCommand | undefined;
  if (!command) usage("缺少命令");
  const options: ParsedCli["options"] = {};
  for (let index = 1; index < argv.length; index += 1) {
    const token = argv[index];
    if (!token.startsWith("--")) usage(`无法识别参数：${token}`);
    const key = token.slice(2);
    if (["force", "json", "dry-run"].includes(key)) {
      options[key] = true;
      continue;
    }
    const value = argv[index + 1];
    if (!value || value.startsWith("--")) usage(`参数 ${token} 缺少值`);
    index += 1;
    if (["done", "takeaway", "problem", "next", "note", "progress"].includes(key)) {
      const current = options[key];
      options[key] = Array.isArray(current) ? [...current, value] : [value];
    } else {
      options[key] = value;
    }
  }
  return { command, options };
}

function usage(message?: string): never {
  if (message) console.error(message);
  console.error("用法：npm run cli -- <init-log|add-log|leetcode|exam-review|dashboard|week-plan|health|doctor|migrate> [options]");
  process.exit(1);
}

function opt(options: ParsedCli["options"], key: string, fallback = ""): string {
  const value = options[key];
  return typeof value === "string" ? value : fallback;
}

function list(options: ParsedCli["options"], key: string): string[] {
  const value = options[key];
  return Array.isArray(value) ? value : [];
}

function requireOpt(options: ParsedCli["options"], key: string): string {
  const value = opt(options, key);
  if (!value) usage(`缺少必填参数 --${key}`);
  return value;
}

function parseRow(value: string, fields: number, label: string): string[] {
  const parts = value.split("|").map((part) => part.trim());
  if (parts.length !== fields) usage(`${label} 需要 ${fields} 个字段，用 | 分隔：${value}`);
  return parts;
}

function logPath(date: string): string {
  return path.join(resolveDataConfig().dataRoot, "logs", `${date}.md`);
}

function renderLog(date: string, status: string, hours: string, goal: string, plan: string): string {
  return `# 学习记录

## 基本信息

- 日期：${date}
- 关联目标：\`${goal}\`
- 关联计划：\`${plan}\`
- 实际学习时长：${hours}
- 今日状态：${status}

## 今日完成

| 任务 | 结果 | 用时 | 证据或产出 |
| --- | --- | ---: | --- |

## 关键收获

## 遇到的问题

| 问题 | 当前判断 | 下一步 |
| --- | --- | --- |

## 明日计划

## 备注
`;
}

function ensureLog(date: string, status: string, hours: string, goal: string, plan: string): string {
  const filePath = logPath(date);
  if (!readable(filePath)) writeText(filePath, ensureRecommendedFrontMatter(renderLog(date, status, hours, goal, plan), relativePath(filePath), date));
  return filePath;
}

function readable(filePath: string): boolean {
  try {
    readText(filePath);
    return true;
  } catch {
    return false;
  }
}

function printRelative(filePath: string): void {
  console.log(relativePath(filePath));
}

function printReport(report: unknown, options: ParsedCli["options"], fallback: string): void {
  if (options.json === true) {
    console.log(JSON.stringify(report, null, 2));
    return;
  }
  console.log(fallback);
}

function cmdInitLog(options: ParsedCli["options"]): void {
  const date = opt(options, "date", today());
  const filePath = ensureLog(
    date,
    opt(options, "status", "顺利"),
    opt(options, "hours"),
    opt(options, "goal", DEFAULT_GOAL),
    opt(options, "plan", DEFAULT_PLAN)
  );
  printRelative(filePath);
}

function cmdAddLog(options: ParsedCli["options"]): void {
  const date = opt(options, "date", today());
  const filePath = ensureLog(
    date,
    opt(options, "status", "顺利"),
    opt(options, "hours"),
    opt(options, "goal", DEFAULT_GOAL),
    opt(options, "plan", DEFAULT_PLAN)
  );
  let text = readText(filePath);
  for (const item of list(options, "done")) {
    text = appendAfterTable(text, "今日完成", markdownRow(parseRow(item, 4, "--done")));
  }
  for (const item of list(options, "problem")) {
    text = appendAfterTable(text, "遇到的问题", markdownRow(parseRow(item, 3, "--problem")));
  }
  for (const item of list(options, "takeaway")) text = appendBullet(text, "关键收获", item);
  for (const item of list(options, "next")) text = appendBullet(text, "明日计划", item);
  for (const item of list(options, "note")) text = appendBullet(text, "备注", item);
  writeText(filePath, text);
  printRelative(filePath);
}

function ensureLeetcodeRecord(): string {
  const filePath = path.join(resolveDataConfig().dataRoot, "records", "leetcode.md");
  if (!readable(filePath)) {
    writeText(
      filePath,
      `# LeetCode 刷题记录

| 日期 | 专题 | 题目 | 难度 | 结果 | 是否重做 | 备注 |
| --- | --- | --- | --- | --- | --- | --- |
`
    );
  }
  return filePath;
}

function cmdLeetcode(options: ParsedCli["options"]): void {
  const filePath = ensureLeetcodeRecord();
  const row = markdownRow([
    opt(options, "date", today()),
    requireOpt(options, "topic"),
    requireOpt(options, "title"),
    opt(options, "difficulty", "中等"),
    opt(options, "result", "待复盘"),
    opt(options, "redo", "否"),
    opt(options, "note")
  ]);
  writeText(filePath, `${readText(filePath).trimEnd()}\n${row}\n`);
  printRelative(filePath);
}

function ensureExamReviewRecord(): string {
  const filePath = path.join(resolveDataConfig().dataRoot, "records", "exam-review.md");
  if (!readable(filePath)) {
    writeText(
      filePath,
      `# 期末复习记录

| 日期 | 科目 | 计划时长 | 实际时长 | 完成内容 | 问题 | 下一步 |
| --- | --- | ---: | ---: | --- | --- | --- |
`
    );
  }
  return filePath;
}

function cmdExamReview(options: ParsedCli["options"]): void {
  const filePath = ensureExamReviewRecord();
  const row = markdownRow([
    opt(options, "date", today()),
    requireOpt(options, "subject"),
    opt(options, "planned"),
    requireOpt(options, "actual"),
    opt(options, "done"),
    opt(options, "problem"),
    opt(options, "next")
  ]);
  writeText(filePath, `${readText(filePath).trimEnd()}\n${row}\n`);
  printRelative(filePath);
}

function replaceFocusLine(text: string, prefix: string, value: string): string {
  if (!value) return text;
  const target = `- ${prefix}：`;
  const lines = text.split(/\r?\n/);
  const index = lines.findIndex((line) => line.startsWith(target));
  if (index === -1) usage(`dashboard.md 中找不到行：${target}`);
  lines[index] = `${target}${value}`;
  return `${lines.join("\n").replace(/\n*$/, "")}\n`;
}

function updateProgressRow(text: string, spec: string, date: string): string {
  const [project, status, stage, milestone] = parseRow(spec, 4, "--progress");
  const lines = text.split(/\r?\n/);
  const existing = lines.findIndex((line) => line.startsWith(`| ${project} |`));
  if (existing >= 0) {
    lines[existing] = markdownRow([project, status, stage, milestone, date]);
    return `${lines.join("\n").replace(/\n*$/, "")}\n`;
  }
  const marker = "| --- | --- | --- | --- | --- |";
  const index = lines.findIndex((line) => line === marker);
  if (index === -1) usage("dashboard.md 中找不到进度概览表");
  lines.splice(index + 1, 0, markdownRow([project, status, stage, milestone, date]));
  return `${lines.join("\n").replace(/\n*$/, "")}\n`;
}

function cmdDashboard(options: ParsedCli["options"]): void {
  const filePath = path.join(resolveDataConfig().dataRoot, "dashboard.md");
  let text = readText(filePath);
  const date = opt(options, "date", today());
  text = replaceFocusLine(text, "当前阶段", opt(options, "stage"));
  text = replaceFocusLine(text, "本周重点", opt(options, "week"));
  text = replaceFocusLine(text, "今日任务", opt(options, "today"));
  for (const item of list(options, "progress")) text = updateProgressRow(text, item, date);
  writeText(filePath, text);
  printRelative(filePath);
}

function cmdWeekPlan(options: ParsedCli["options"]): void {
  const now = new Date();
  const week = opt(options, "week", `${now.getFullYear()}-W${String(weekNumber(now)).padStart(2, "0")}`);
  const filePath = path.join(resolveDataConfig().dataRoot, "plans", `${week}.md`);
  if (readable(filePath) && options.force !== true) {
    usage(`${relativePath(filePath)} 已存在；如需覆盖请加 --force`);
  }
  const text = `# 学习计划：${week}

## 基本信息

- 计划周期：${week}
- 关联目标：\`${opt(options, "goal", DEFAULT_GOAL)}\`
- 关联路线：\`${opt(options, "route", "routes/backend-plus-ai-application-roadmap.md")}\`
- 本周主题：${opt(options, "theme", "待填写")}
- 计划总时长：${opt(options, "hours", "待填写")}

## 本周期目标

| 目标 | 验收方式 | 优先级 |
| --- | --- | --- |

## 任务安排

| 日期 | 任务 | 预计用时 | 产出物 | 状态 |
| --- | --- | ---: | --- | --- |

## 每日最低动作

- 刷 1 道算法题，并记录解法要点。
- 学习 30 分钟后端、计算机基础或 Linux 实操。
- 在 \`logs/\` 中记录当天完成内容、问题和下一步。

## 重点问题

| 问题 | 解决路径 | 截止时间 |
| --- | --- | --- |

## 复盘入口

- 对应复盘文件：\`reviews/${week}.md\`
`;
  writeText(filePath, ensureRecommendedFrontMatter(text || defaultContent("plans", week), relativePath(filePath), week));
  printRelative(filePath);
}

function cmdHealth(options: ParsedCli["options"]): void {
  const report = healthReport();
  printReport(report, options, `health ok=${report.ok} schema=${report.schema_version} issues=${report.issues.length}`);
}

function cmdDoctor(options: ParsedCli["options"]): void {
  const report = doctorReport();
  printReport(report, options, `doctor ok=${report.ok} schema=${report.schema_version} checks=${report.checks.length} issues=${report.health.issues.length}`);
}

function cmdMigrate(options: ParsedCli["options"]): void {
  const report = migrateWorkspace({ dryRun: options["dry-run"] === true });
  printReport(report, options, `migrate ok=${report.ok} dry_run=${report.dry_run} actions=${report.actions.length} backups=${report.backups.length}`);
}

function weekNumber(date: Date): number {
  const target = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
  const day = target.getUTCDay() || 7;
  target.setUTCDate(target.getUTCDate() + 4 - day);
  const yearStart = new Date(Date.UTC(target.getUTCFullYear(), 0, 1));
  return Math.ceil(((target.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
}

const parsed = parseArgs(process.argv.slice(2));

switch (parsed.command) {
  case "init-log":
    cmdInitLog(parsed.options);
    break;
  case "add-log":
    cmdAddLog(parsed.options);
    break;
  case "leetcode":
    cmdLeetcode(parsed.options);
    break;
  case "exam-review":
    cmdExamReview(parsed.options);
    break;
  case "dashboard":
    cmdDashboard(parsed.options);
    break;
  case "week-plan":
    cmdWeekPlan(parsed.options);
    break;
  case "health":
    cmdHealth(parsed.options);
    break;
  case "doctor":
    cmdDoctor(parsed.options);
    break;
  case "migrate":
    cmdMigrate(parsed.options);
    break;
  default:
    usage(`未知命令：${parsed.command}`);
}
